shared/p2putils/fork.go

```diff
@@ -43,24 +82,83 @@ func CreateForkDigest(
 // returns the active fork version during this epoch.
 func Fork(
 	targetEpoch types.Epoch,
-) (*statepb.Fork, error) {
-	// We retrieve a list of scheduled forks by epoch.
-	// We loop through the keys in this map to determine the current
-	// fork version based on the requested epoch.
-	retrievedForkVersion := params.BeaconConfig().GenesisForkVersion
-	previousForkVersion := params.BeaconConfig().GenesisForkVersion
-	scheduledForks := params.BeaconConfig().ForkVersionSchedule
+) (*ethpb.Fork, error) {
+	currentForkVersion := bytesutil.ToBytes4(params.BeaconConfig().GenesisForkVersion)
+	previousForkVersion := bytesutil.ToBytes4(params.BeaconConfig().GenesisForkVersion)
+	fSchedule := params.BeaconConfig().ForkVersionSchedule
+	sortedForkVersions := SortedForkVersions(fSchedule)
 	forkEpoch := types.Epoch(0)
-	for epoch, forkVersion := range scheduledForks {
-		if epoch <= targetEpoch {
-			previousForkVersion = retrievedForkVersion
-			retrievedForkVersion = forkVersion
+	for _, forkVersion := range sortedForkVersions {
+		epoch := fSchedule[forkVersion]
+		if targetEpoch >= epoch {
+			previousForkVersion = currentForkVersion
+			currentForkVersion = forkVersion
 			forkEpoch = epoch
 		}
 	}
-	return &statepb.Fork{
-		PreviousVersion: previousForkVersion,
-		CurrentVersion:  retrievedForkVersion,
+	return &ethpb.Fork{
+		PreviousVersion: previousForkVersion[:],
+		CurrentVersion:  currentForkVersion[:],
 		Epoch:           forkEpoch,
 	}, nil
 }
+
+// RetrieveForkDataFromDigest performs the inverse, where it tries to determine the fork version
+// and epoch from a provided digest by looping through our current fork schedule.
+func RetrieveForkDataFromDigest(digest [4]byte, genesisValidatorsRoot []byte) ([4]byte, types.Epoch, error) {
+	fSchedule := params.BeaconConfig().ForkVersionSchedule
+	for v, e := range fSchedule {
+		rDigest, err := helpers.ComputeForkDigest(v[:], genesisValidatorsRoot)
+		if err != nil {
+			return [4]byte{}, 0, err
+		}
+		if rDigest == digest {
+			return v, e, nil
+		}
+	}
+	return [4]byte{}, 0, errors.Errorf("no fork exists for a digest of %#x", digest)
+}
+
+// NextForkData retrieves the next fork data according to the
+// provided current epoch.
+func NextForkData(currEpoch types.Epoch) ([4]byte, types.Epoch) {
+	fSchedule := params.BeaconConfig().ForkVersionSchedule
+	sortedForkVersions := SortedForkVersions(fSchedule)
+	nextForkEpoch := types.Epoch(math.MaxUint64)
+	nextForkVersion := [4]byte{}
+	for _, forkVersion := range sortedForkVersions {
+		epoch := fSchedule[forkVersion]
```

Comment:
I wonder if we should be safer here and check for existence in the map with `ok`. Same feedback for different accessors of this map